extensions/positron-r/src/session.ts

```diff
@@ -385,71 +388,155 @@ export class RSession implements positron.LanguageRuntimeSession, vscode.Disposa
+		return pkgInst;
+	}
 
-				// Wait for the the runtime to be idle, or for the timeout:
-				await Promise.race([promise, timeout(2e4, 'waiting for package installation')]);
+	/**
+	 * Checks whether a package is installed in the runtime, possibly at a minimum version. If not,
+	 * prompts the user to install the package. See the documentation for `packageVersion() for some
+	 * caveats around caching.
+	 * @param pkgName The name of the package to check.
+	 * @param minimumVersion Optionally, the version of the package needed.
+	 * @returns true if the package is installed, at a sufficient version, false otherwise.
+	 */
 
-				return true;
-			} else {
-				return false;
-			}
+	async checkInstalled(pkgName: string, minimumVersion: string | null = null): Promise<boolean> {
+		let pkgInst = await this.packageVersion(pkgName, minimumVersion);
+		const installed = pkgInst !== null;
+		let compatible = pkgInst?.compatible ?? false;
+		if (compatible) {
+			return true;
 		}
-		this._packageCache.push({ packageName: pkgName, packageVersion: pkgVersion });
-		return true;
+		// One of these is true:
+		// - Package is not installed.
+		// - Package is installed, but version is insufficient.
+		// - (Our cache gave us outdated info, but we're just accepting this risk.)
+
+		const title = installed
+			? vscode.l10n.t('Insufficient package version')
+			: vscode.l10n.t('Missing R package');
+		const message = installed
+			? vscode.l10n.t(
+				'The {0} package is installed at version {1}, but version {2} is required.',
+				pkgName, pkgInst!.packageVersion, minimumVersion as string
+			)
+			: vscode.l10n.t('The {0} package is required, but not installed.', pkgName);
+		const okButtonTitle = installed
+			? vscode.l10n.t('Update now')
+			: vscode.l10n.t('Install now');
+
+		const install = await positron.window.showSimpleModalDialogPrompt(
+			title,
+			message,
+			okButtonTitle
+		);
+		if (!install) {
+			return false;
+		}
+
+		const id = randomUUID();
+
+		// A promise that resolves when the runtime is idle:
+		const promise = new Promise<void>(resolve => {
+			const disp = this.onDidReceiveRuntimeMessage(runtimeMessage => {
+				if (runtimeMessage.parent_id === id &&
+					runtimeMessage.type === positron.LanguageRuntimeMessageType.State) {
+					const runtimeMessageState = runtimeMessage as positron.LanguageRuntimeState;
+					if (runtimeMessageState.state === positron.RuntimeOnlineState.Idle) {
+						resolve();
+						disp.dispose();
+					}
+				}
+			});
+		});
+
+		this.execute(`install.packages("${pkgName}")`,
+			id,
+			positron.RuntimeCodeExecutionMode.Interactive,
+			positron.RuntimeErrorBehavior.Continue);
+
+		// Wait for the the runtime to be idle, or for the timeout:
+		await Promise.race([promise, timeout(2e4, 'waiting for package installation')]);
+
+		pkgInst = await this.packageVersion(pkgName, minimumVersion);
+		compatible = pkgInst?.compatible ?? false;
+		return compatible;
```

Comment:
This is a bit new: previously we always assumed that our attempt to install/update the package was successful. Now we check that specifically, cache new version info, and return the result (hopefully `true`).